### ISSUE TITLE
RDT: non-overlapping cache allocations

### ISSUE THREAD

**Issue by marquiz** [intel/cri-resource-manager#35]
Currently, there is no way of specifying non-overlapping L3 schemas using the relative (percentage) notation. We might want to specify something like this (using the percentage notation):
```
L3:          XXXXXXXXXXXXXXXXXXXX
Guaranteed:  xxxxxxxxxx
Burstable              xxxxxxxxxx
Besteffort                  xxxxx
```

For this we would need both exclusive non-overlapping, and, non-exclusive overlapping definitions at the same time. One way to achieve this would be to add one extra level of hierarchy (i.e. groups) to the class configuration. Cache resources could not overlap on the group level. The scheme described above could be achieved with a configuration something like this:
```
resctrlGroups:
  exclusive:
    l3allocation: "50%"
    classes:
      Guaranteed:
        l3schema:
          default: "100%"
  shared:
    l3allocation: "50%"
    classes:
      Burstable:
        l3schema:
          default: "100%"
      BestEffort:
        l3schema:
          default: "50%"
```
In this scheme, Memory Bandwidth (MB) would probably allow over-committing i.e. each group could have e.g. 100%.

**Comment by kad** [comment]
Exclusive cache allocation is something that can be tricky. User might want to have multiple workloads each of those which should have exclusive portion of cache. With just one more layer in hierarchy it would mean that exclusive portion will be shared with all containers that are part of this "exclusive" class. Maybe this is not bad in overall, but maybe we can think about more flexibility here.

**Comment by linericyang** [comment]
Allowing admin to set RDT resctrl bitmask directly will provide the most flexibility. However, it requires admin has knowledge about resctrl bit mask format and hardware platform details, like how many cache ways available on each socket.

```
Guaranteed:
  l3CacheSchema: L3:0=fff0;1=fff00
  memBwSchema: MB:0=20;1=70
Burstable:
  l3CacheSchema: L3:0=000ff;1=000ff
  memBwSchema: MB:0=20;1=70
```
Or using percentage notation, it still can be used to statify overlapping or non-overlapping case.
```
Guaranteed:
  l3Cache: 0% - 60%
Burstable:
  l3Cache: 60% - 100%
BestEffort:
  l3Cache: 80% - 100%
```
@marquiz May I ask what is the status of this feature? I would like to help if you need.

**Comment by marquiz** [comment]
> @marquiz May I ask what is the status of this feature? I would like to help if you need.

It is implemented in #74. The idea is to define "buckets" which are always exclusive wrt. to each other. Groups  inside buckets share the resources allocated for that specific bucket. I think it would be ready for merging but you're very welcome to comment/review that.



> Allowing admin to set RDT resctrl bitmask directly will provide the most flexibility. However, it requires admin has knowledge about resctrl bit mask format and hardware platform details, like how many cache ways available on each socket.

That is possible of course – already now the user can specify explicit bitmasks and achieve exclusive allocation through that. The problem is, as you said, that it requires detailed knowledge of the underlying hardware and does not suit that well in heterogenous clusters where nodes may be different.



> Or using percentage notation, it still can be used to statify overlapping or non-overlapping case.

It is not quite as simple as that as the number of available bits in the bitmask is (significantly) lower than 100. E.g. imagine a situation where there are 10 bits available. In this case 0-5% and 5-10% both map to the same bit. Also, the minimum number of bits that must be allocated for each group can be greater than one. Thus, the heuristics of resolving the allocation (and trying to guess what the user implied) can get really tricky. That's why I implemented the the non-overlapping allocation using the buckets in #74

**Comment by marquiz** [comment]
Implemented in #74
